--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/common/cmm_wpa.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/common/cmm_wpa.c
@@ -1227,7 +1227,7 @@
 	GenRandom(pAd, (UCHAR *)pCurrentAddr, pEntry->SNonce);
 	pEntry->AllowInsPTK = TRUE;
 	pEntry->LastGroupKeyId = 0;
-	NdisZeroMemory(pEntry->LastGTK, 32);
+	NdisZeroMemory(pEntry->LastGTK, MAX_LEN_GTK);
 
 #ifdef DOT11R_FT_SUPPORT	
 	if (IS_FT_RSN_STA(pEntry))
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/common/cmm_wpa.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/common/cmm_wpa.c
@@ -1111,7 +1111,7 @@
 	GenRandom(pAd, (UCHAR *)pCurrentAddr, pEntry->SNonce);
 	pEntry->AllowInsPTK = TRUE;
 	pEntry->LastGroupKeyId = 0;
-	NdisZeroMemory(pEntry->LastGTK, 32);
+	NdisZeroMemory(pEntry->LastGTK, MAX_LEN_GTK);
 
 #ifdef DOT11R_FT_SUPPORT
 	if (IS_FT_RSN_STA(pEntry))
