--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_connect.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_connect.c
@@ -302,8 +302,6 @@
 	}
 #endif /* DOT11R_FT_SUPPORT */
 
-	BeaconTransmit.word = 0;
-
 #ifdef SPECIFIC_TX_POWER_SUPPORT
         /* Specific Power for Long-Range Beacon */
 	if ((pAd->ApCfg.MBSSID[apidx].TxPwrAdj != -1) /* && 
@@ -312,6 +310,8 @@
 		TxPwrAdj = pAd->ApCfg.MBSSID[apidx].TxPwrAdj;
 	}
 #endif /* SPECIFIC_TX_POWER_SUPPORT */
+
+	BeaconTransmit.word = 0;
 
 	RTMPWriteTxWI(pAd, &pAd->BeaconTxWI, FALSE, FALSE, TRUE, FALSE, FALSE, TRUE, 0, BSS0Mcast_WCID, 
 					FrameLen, PID_MGMT, 0, 0,IFS_HTTXOP, &BeaconTransmit);
