--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_autoChSel.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_autoChSel.c
@@ -1081,13 +1081,13 @@
 		if (pBssInfoTab->BssNr >= MAX_LEN_OF_BSS_TABLE)
 			return BSS_NOT_FOUND;
 		Idx = pBssInfoTab->BssNr;
-		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx], pBssid, Ssid, SsidLen,
+		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx % MAX_LEN_OF_BSS_TABLE], pBssid, Ssid, SsidLen,
 							ChannelNo, ExtChOffset, Rssi);
 		pBssInfoTab->BssNr++;
 	} 
 	else
 	{
-		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx], pBssid, Ssid, SsidLen,
+		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx % MAX_LEN_OF_BSS_TABLE], pBssid, Ssid, SsidLen,
 							ChannelNo, ExtChOffset, Rssi);
 	}
 
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/oid.h
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/oid.h
@@ -77,17 +77,6 @@
 #define MAX_NUMBER_OF_ACL			64
 #define MAX_LENGTH_OF_SUPPORT_RATES		12	/* 1, 2, 5.5, 11, 6, 9, 12, 18, 24, 36, 48, 54 */
 #define MAX_NUMBER_OF_DLS_ENTRY			4
-
-
-#ifdef CUSTOMER_DCC_FEATURE
-#ifdef MEMORY_OPTIMIZATION
-#define MAX_LEN_OF_BSS_TABLE			1
-#define MAX_REORDERING_MPDU_NUM			256
-#else
-#define MAX_LEN_OF_BSS_TABLE			64
-#define MAX_REORDERING_MPDU_NUM			512
-#endif
-#endif
 
 #define RT_QUERY_SIGNAL_CONTEXT				0x0402
 #define RT_SET_IAPP_PID                 	0x0404
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/rtmp_def.h
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/rtmp_def.h
@@ -633,7 +633,6 @@
 
 #define MAX_LEN_OF_BA_REC_TABLE          ((NUM_OF_TID * MAX_LEN_OF_MAC_TABLE)/2)	/*   (NUM_OF_TID*MAX_AID_BA + 32)        //Block ACK recipient */
 #define MAX_LEN_OF_BA_ORI_TABLE          ((NUM_OF_TID * MAX_LEN_OF_MAC_TABLE)/2)	/*   (NUM_OF_TID*MAX_AID_BA + 32)   // Block ACK originator */
-#ifndef CUSTOMER_DCC_FEATURE
 #ifdef MEMORY_OPTIMIZATION
 #define MAX_LEN_OF_BSS_TABLE             1
 #define MAX_REORDERING_MPDU_NUM			 256
@@ -641,7 +640,7 @@
 #define MAX_LEN_OF_BSS_TABLE             128 /* 64 */
 #define MAX_REORDERING_MPDU_NUM			 512
 #endif
-#endif
+
 /* key related definitions */
 #define SHARE_KEY_NUM                   4
 #define MAX_LEN_OF_SHARE_KEY            16	/* byte count */
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_autoChSel.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_autoChSel.c
@@ -959,13 +959,13 @@
 		if (pBssInfoTab->BssNr >= MAX_LEN_OF_BSS_TABLE)
 			return BSS_NOT_FOUND;
 		Idx = pBssInfoTab->BssNr;
-		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx], pBssid, Ssid, SsidLen,
+		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx % MAX_LEN_OF_BSS_TABLE], pBssid, Ssid, SsidLen,
 							ChannelNo, ExtChOffset, Rssi);
 		pBssInfoTab->BssNr++;
 	} 
 	else
 	{
-		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx], pBssid, Ssid, SsidLen,
+		AutoChBssEntrySet(&pBssInfoTab->BssEntry[Idx % MAX_LEN_OF_BSS_TABLE], pBssid, Ssid, SsidLen,
 							ChannelNo, ExtChOffset, Rssi);
 	}
 
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/include/rtmp_def.h
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/include/rtmp_def.h
@@ -671,15 +671,7 @@
 #define MAX_LEN_OF_BSS_TABLE             1
 #define MAX_REORDERING_MPDU_NUM			 256
 #else
-#ifdef BB_SOC
-#define MAX_LEN_OF_BSS_TABLE             20
-#else /* BB_SOC */
-#ifdef ECONET_ALPHA_RELEASE
-#define MAX_LEN_OF_BSS_TABLE             96 /* 64 */
-#else
 #define MAX_LEN_OF_BSS_TABLE             128 /* 64 */
-#endif /* ECONET_ALPHA_RELEASE */
-#endif /* !BB_SOC */
 #define MAX_REORDERING_MPDU_NUM			 512
 #endif
 
