--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
@@ -898,6 +898,9 @@
     
 	/* clear the previous Pairwise key table */
     if(pEntry->Aid != 0 &&
+#ifdef DOT11R_FT_SUPPORT
+	(!IS_FT_STA(pEntry)) &&
+#endif
 	(pEntry->WepStatus >= Ndis802_11TKIPEnable 
 #ifdef DOT1X_SUPPORT
 	|| wdev->IEEE8021X
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_assoc.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_assoc.c
@@ -1276,6 +1276,9 @@
     
 	/* clear the previous Pairwise key table */
     if(pEntry->Aid != 0 &&
+#ifdef DOT11R_FT_SUPPORT
+	(!IS_FT_STA(pEntry)) &&
+#endif
 	(pEntry->WepStatus >= Ndis802_11TKIPEnable 
 #ifdef DOT1X_SUPPORT
 	|| wdev->IEEE8021X
