--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
@@ -823,6 +823,9 @@
 	BOOLEAN bAllowStaConnectInHt = FALSE;
 #endif /* BAND_STEERING */
 	CHAR rssi;
+#ifdef DOT11R_FT_SUPPORT
+	UCHAR zeroFT[LEN_TK];
+#endif
 #ifdef RT_BIG_ENDIAN
 	UINT32 tmp_1;
 	UINT64 tmp_2;
@@ -895,11 +898,15 @@
                 goto SendAssocResponse;
         }
 #endif /* DOT11W_PMF_SUPPORT */
-    
-	/* clear the previous Pairwise key table */
+
+#ifdef DOT11R_FT_SUPPORT
+    NdisZeroMemory(zeroFT, LEN_TK);
+#endif
+
+    /* clear the previous Pairwise key table */
     if(pEntry->Aid != 0 &&
 #ifdef DOT11R_FT_SUPPORT
-	(!IS_FT_STA(pEntry)) &&
+	(pEntry->AllowInsPTK == TRUE || !IS_FT_STA(pEntry) || (IS_FT_STA(pEntry) && (NdisEqualMemory(&pEntry->PTK[OFFSET_OF_PTK_TK], zeroFT, LEN_TK) || !NdisEqualMemory(&pEntry->PTK[OFFSET_OF_PTK_TK], pEntry->LastTK, LEN_TK)))) &&
 #endif
 	(pEntry->WepStatus >= Ndis802_11TKIPEnable 
 #ifdef DOT1X_SUPPORT
@@ -1342,8 +1349,13 @@
 			NdisMoveMemory(pFtIe->MIC, ft_mic, FT_MIC_LEN);
 
 			/* Only first allow install from assoc, later or rekey or install from auth (backward compatability with not patched clients) */
-			if (pEntry->AllowInsPTK == TRUE) {
+			if(pEntry->AllowInsPTK == TRUE
+#ifdef DOT11R_FT_SUPPORT
+			    || !IS_FT_STA(pEntry) || (IS_FT_STA(pEntry) && (NdisEqualMemory(&pEntry->PTK[OFFSET_OF_PTK_TK], zeroFT, LEN_TK) || !NdisEqualMemory(&pEntry->PTK[OFFSET_OF_PTK_TK], pEntry->LastTK, LEN_TK)))
+#endif
+			) {
 			    WPAInstallPairwiseKey(pAd, pEntry->apidx, pEntry, TRUE);
+			    NdisMoveMemory(pEntry->LastTK, &pEntry->PTK[OFFSET_OF_PTK_TK], LEN_TK);
 			    pEntry->AllowInsPTK = FALSE;
 			}
 
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_auth.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_auth.c
@@ -583,6 +583,7 @@
     						&pFtInfoBuf->MdIeInfo, &pFtInfoBuf->FtIeInfo, NULL,
     						pFtInfoBuf->RSN_IE, pFtInfoBuf->RSNIE_Len);
 
+				NdisZeroMemory(pEntry->LastTK, LEN_TK);
 				os_free_mem(NULL, pFtInfoBuf);
 				if (result == MLME_SUCCESS) {
 					/* Install pairwise key */
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/rtmp.h
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/include/rtmp.h
@@ -3022,6 +3022,7 @@
 	BOOLEAN AllowInsPTK;
 	UCHAR LastGroupKeyId;
 	UCHAR LastGTK[MAX_LEN_GTK];
+	UCHAR LastTK[LEN_TK];
 	NDIS_802_11_PRIVACY_FILTER PrivacyFilter;	/* PrivacyFilter enum for 802.1X */
 	CIPHER_KEY PairwiseKey;
 	INT PMKID_CacheIdx;
