--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_assoc.c
@@ -889,7 +889,14 @@
 			break;
 		}
 	}
-    
+
+	/* drop this assoc req by silencely discard this frame */
+	if (FlgIs11bSta == 1 && (pAd->LatchRfRegs.Channel > 14 || !WMODE_EQUAL(PhyMode, WMODE_B))) {
+		DBGPRINT(RT_DEBUG_ERROR, ("%s():pEntry is 11b only STA and AP do not support 11B rates\n",
+					__FUNCTION__));
+		goto LabelOK;
+	}
+
 #ifdef DOT11W_PMF_SUPPORT
         if ((pEntry->PortSecured == WPA_802_1X_PORT_SECURED)
                 && (CLIENT_STATUS_TEST_FLAG(pEntry, fCLIENT_STATUS_PMF_CAPABLE)))
@@ -1040,6 +1047,12 @@
 	/* 2. qualify this STA's auth_asoc status in the MAC table, decide StatusCode */
 	StatusCode = APBuildAssociation(pAd, pEntry, ie_list, MaxSupportedRate, &Aid);
 
+	/* drop this assoc req and send reject */
+	if (StatusCode == MLME_ASSOC_REJ_DATA_RATE || StatusCode == MLME_UNSPECIFY_FAIL) {
+		//RTMPSendWirelessEvent(pAd, IW_STA_MODE_EVENT_FLAG, pEntry->Addr, pEntry->apidx, 0);
+		goto SendAssocResponse;
+	}
+
 #ifdef DOT11R_FT_SUPPORT
 	if (pEntry->apidx < pAd->ApCfg.BssidNum)
 	{
@@ -1079,12 +1092,8 @@
 	}
 #endif /* DOT11_VHT_AC */
 
-	if (StatusCode == MLME_ASSOC_REJ_DATA_RATE)
-		RTMPSendWirelessEvent(pAd, IW_STA_MODE_EVENT_FLAG, pEntry->Addr, pEntry->apidx, 0);
-
-#ifdef DOT11W_PMF_SUPPORT
 SendAssocResponse:
-#endif /* DOT11W_PMF_SUPPORT */
+
 	/* 3. send Association Response */
 	NStatus = MlmeAllocateMemory(pAd, &pOutBuffer);
 	if (NStatus != NDIS_STATUS_SUCCESS)
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_assoc.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_assoc.c
@@ -1277,7 +1277,14 @@
 			break;
 		}
 	}
-    
+
+	/* drop this assoc req by silencely discard this frame */
+	if (FlgIs11bSta == 1 && (pAd->LatchRfRegs.Channel > 14 || !WMODE_EQUAL(PhyMode, WMODE_B))) {
+		DBGPRINT(RT_DEBUG_ERROR, ("%s():pEntry is 11b only STA and AP do not support 11B rates\n",
+					__FUNCTION__));
+		goto LabelOK;
+	}
+
 #ifdef DOT11W_PMF_SUPPORT
         if ((tr_entry->PortSecured == WPA_802_1X_PORT_SECURED)
                 && (CLIENT_STATUS_TEST_FLAG(pEntry, fCLIENT_STATUS_PMF_CAPABLE)))
@@ -1439,6 +1446,12 @@
 	/* 2. qualify this STA's auth_asoc status in the MAC table, decide StatusCode */
 	StatusCode = APBuildAssociation(pAd, pEntry, ie_list, MaxSupportedRate, &Aid);
 
+	/* drop this assoc req and send reject */
+	if (StatusCode == MLME_ASSOC_REJ_DATA_RATE || StatusCode == MLME_UNSPECIFY_FAIL) {
+		//RTMPSendWirelessEvent(pAd, IW_STA_MODE_EVENT_FLAG, pEntry->Addr, wdev->wdev_idx, 0);
+		goto SendAssocResponse;
+	}
+
 #ifdef DOT11R_FT_SUPPORT
 	if (pEntry->func_tb_idx < pAd->ApCfg.BssidNum)
 	{
@@ -1478,12 +1491,8 @@
 	}
 #endif /* DOT11_VHT_AC */
 
-	if (StatusCode == MLME_ASSOC_REJ_DATA_RATE)
-		RTMPSendWirelessEvent(pAd, IW_STA_MODE_EVENT_FLAG, pEntry->Addr, wdev->wdev_idx, 0);
-
-#ifdef DOT11W_PMF_SUPPORT
 SendAssocResponse:
-#endif /* DOT11W_PMF_SUPPORT */
+
 	/* 3. send Association Response */
 	NStatus = MlmeAllocateMemory(pAd, &pOutBuffer);
 	if (NStatus != NDIS_STATUS_SUCCESS)
