--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_data.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/3.0.X.X/mt76x2/ap/ap_data.c
@@ -5120,14 +5120,20 @@
 #endif /* MWDS */
 #ifdef WDS_SUPPORT
 			/* handle WDS */
-			{
+			if (!pEntry)
+			{
+				/*
+					The WDS frame only can go here when in auto learning mode and
+					this is the first trigger frame from peer
+
+					So we check if this is un-registered WDS entry by call function
+						"FindWdsEntry()"
+				*/
 				if (MAC_ADDR_EQUAL(pHeader->Addr1, pAd->CurrentAddress))
 					pEntry = FindWdsEntry(pAd, pRxBlk->wcid, pHeader->Addr2, pRxBlk->rx_rate.field.MODE);
-				else
-					pEntry = NULL;
 
 				/* have no valid wds entry exist, then discard the incoming packet.*/
-				if (!(pEntry && WDS_IF_UP_CHECK(pAd, pEntry->wdev_idx)))
+				if (!pEntry || !WDS_IF_UP_CHECK(pAd, pEntry->wdev_idx))
 					goto err;
 
 				/*receive corresponding WDS packet, disable TX lock state (fix WDS jam issue) */
@@ -5744,20 +5750,23 @@
 #endif /* MWDS */
 #ifdef WDS_SUPPORT
 			/* handle WDS */
-			{
+			if (!pEntry)
+			{
+				/*
+					The WDS frame only can go here when in auto learning mode and
+					this is the first trigger frame from peer
+
+					So we check if this is un-registered WDS entry by call function
+						"FindWdsEntry()"
+				*/
 				bWdsPacket = TRUE;
 				if (MAC_ADDR_EQUAL(pHeader->Addr1, pAd->CurrentAddress))
 					pEntry = FindWdsEntry(pAd, pRxBlk->wcid, pHeader->Addr2, pRxBlk->rx_rate.field.MODE);
-				else
-					pEntry = NULL;
 
 
 				/* have no valid wds entry exist, then discard the incoming packet.*/
-				if (!(pEntry && WDS_IF_UP_CHECK(pAd, pEntry->wdev_idx)))
-				{
-					/* drop the packet */
+				if (!pEntry  || !WDS_IF_UP_CHECK(pAd, pEntry->wdev_idx))
 					goto err;
-				}
 
 				/*receive corresponding WDS packet, disable TX lock state (fix WDS jam issue) */
 				if(pEntry && (pEntry->LockEntryTx == TRUE)) 
@@ -5766,7 +5775,7 @@
 					pEntry->ContinueTxFailCnt = 0;
 					pEntry->LockEntryTx = FALSE;
 				}
-		
+
 				RX_BLK_SET_FLAG(pRxBlk, fRX_WDS);
 				FromWhichBSSID = pEntry->wdev_idx + MIN_NET_DEVICE_FOR_WDS;
 				break;
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_data.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_data.c
@@ -3462,7 +3462,7 @@
 				pEntry = FindWdsEntry(pAd, pRxBlk->wcid, pHeader->Addr2, pRxBlk->rx_rate.field.MODE);
 
 			/* have no valid wds entry exist, then discard the incoming packet.*/
-			if (!(pEntry && WDS_IF_UP_CHECK(pAd, pEntry->func_tb_idx)))
+			if (!pEntry || !WDS_IF_UP_CHECK(pAd, pEntry->func_tb_idx))
 				return FALSE;
 
 			/*receive corresponding WDS packet, disable TX lock state (fix WDS jam issue) */
--- rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_wds.c
+++ rt-n56u/trunk/proprietary/rt_wifi/rtpci/4.1.X.X/mt76x3/ap/ap_wds.c
@@ -90,7 +90,7 @@
 			pEntry = NULL;
 
 		/* have no valid wds entry exist, then discard the incoming packet.*/
-		if (!(pEntry && WDS_IF_UP_CHECK(pAd, pEntry->func_tb_idx)))
+		if (!pEntry || !WDS_IF_UP_CHECK(pAd, pEntry->func_tb_idx))
 			return FALSE;
 
 		/*receive corresponding WDS packet, disable TX lock state (fix WDS jam issue) */
