diff -BurpN ppp.2.4.5/pppd/ipv6cp.c ppp/pppd/ipv6cp.c
--- ppp.2.4.5/pppd/ipv6cp.c	2005-11-04 10:20:14.000000000 +0000
+++ ppp/pppd/ipv6cp.c	2010-10-23 13:36:07.530000002 +0000
@@ -176,6 +176,8 @@ ipv6cp_options ipv6cp_gotoptions[NUM_PPP
 ipv6cp_options ipv6cp_allowoptions[NUM_PPP];	/* Options we allow peer to request */
 ipv6cp_options ipv6cp_hisoptions[NUM_PPP];	/* Options that we ack'd */
 int no_ifaceid_neg = 0;
+char path_ipv6up[MAXPATHLEN];			/* pathname of ipv6-up script */
+char path_ipv6down[MAXPATHLEN];			/* pathname of ipv6-down script */
 
 /* local vars */
 static int ipv6cp_is_up;
@@ -255,6 +257,13 @@ static option_t ipv6cp_option_list[] = {
     { "ipv6cp-max-failure", o_int, &ipv6cp_fsm[0].maxnakloops,
       "Set max #conf-naks for IPv6CP", OPT_PRIO },
 
+    { "ipv6-up-script", o_string, path_ipv6up,
+      "Set pathname of ipv6-up script",
+      OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },
+    { "ipv6-down-script", o_string, path_ipv6down,
+      "Set pathname of ipv6-down script",
+      OPT_PRIV|OPT_STATIC, NULL, MAXPATHLEN },
+
    { NULL }
 };
 
@@ -1294,7 +1303,7 @@ ipv6cp_up(f)
      */
     if (ipv6cp_script_state == s_down && ipv6cp_script_pid == 0) {
 	ipv6cp_script_state = s_up;
-	ipv6cp_script(_PATH_IPV6UP);
+	ipv6cp_script(path_ipv6up);
     }
 }
 
@@ -1345,7 +1354,7 @@ ipv6cp_down(f)
     /* Execute the ipv6-down script */
     if (ipv6cp_script_state == s_up && ipv6cp_script_pid == 0) {
 	ipv6cp_script_state = s_down;
-	ipv6cp_script(_PATH_IPV6DOWN);
+	ipv6cp_script(path_ipv6down);
     }
 }
 
@@ -1388,13 +1397,13 @@ ipv6cp_script_done(arg)
     case s_up:
 	if (ipv6cp_fsm[0].state != OPENED) {
 	    ipv6cp_script_state = s_down;
-	    ipv6cp_script(_PATH_IPV6DOWN);
+	    ipv6cp_script(path_ipv6down);
 	}
 	break;
     case s_down:
 	if (ipv6cp_fsm[0].state == OPENED) {
 	    ipv6cp_script_state = s_up;
-	    ipv6cp_script(_PATH_IPV6UP);
+	    ipv6cp_script(path_ipv6up);
 	}
 	break;
     }
diff -BurpN ppp.2.4.5/pppd/ipv6cp.h ppp/pppd/ipv6cp.h
--- ppp.2.4.5/pppd/ipv6cp.h	2005-11-04 10:20:55.000000000 +0000
+++ ppp/pppd/ipv6cp.h	2010-10-23 13:36:41.630000002 +0000
@@ -170,3 +170,6 @@ extern ipv6cp_options ipv6cp_allowoption
 extern ipv6cp_options ipv6cp_hisoptions[];
 
 extern struct protent ipv6cp_protent;
+
+extern char path_ipv6up[MAXPATHLEN];	/* pathname of ipv6-up script */
+extern char path_ipv6down[MAXPATHLEN];	/* pathname of ipv6-down script */
diff -BurpN ppp.2.4.5/pppd/main.c ppp/pppd/main.c
--- ppp.2.4.5/pppd/main.c	2009-05-22 09:15:10.000000000 +0000
+++ ppp/pppd/main.c	2010-10-23 13:38:38.780000002 +0000
@@ -320,6 +320,10 @@ main(argc, argv)
 
     strlcpy(path_ipup, _PATH_IPUP, sizeof(path_ipup));
     strlcpy(path_ipdown, _PATH_IPDOWN, sizeof(path_ipdown));
+#ifdef INET6
+    strlcpy(path_ipv6up, _PATH_IPV6UP, sizeof(path_ipv6up));
+    strlcpy(path_ipv6down, _PATH_IPV6DOWN, sizeof(path_ipv6down));
+#endif
 
     link_stats_valid = 0;
     new_phase(PHASE_INITIALIZE);
