diff -ur transmission-2.82/libtransmission/rpc-server.c transmission-2.8x/libtransmission/rpc-server.c
--- transmission-2.82/libtransmission/rpc-server.c	2013-08-09 10:45:45.000000000 +0800
+++ transmission-2.8x/libtransmission/rpc-server.c	2013-08-09 23:51:59.871380634 +0800
@@ -364,11 +364,7 @@
 
           /* zlib's manual says: "Add 16 to windowBits to write a simple gzip header
            * and trailer around the compressed data instead of a zlib wrapper." */
-#ifdef TR_LIGHTWEIGHT
           compressionLevel = Z_DEFAULT_COMPRESSION;
-#else
-          compressionLevel = Z_BEST_COMPRESSION;
-#endif
           deflateInit2 (&server->stream, compressionLevel, Z_DEFLATED, 15+16, 8, Z_DEFAULT_STRATEGY);
         }
 
diff -ur transmission-2.82/libtransmission/tr-udp.c transmission-2.8x/libtransmission/tr-udp.c
--- transmission-2.82/libtransmission/tr-udp.c	2013-08-09 10:45:41.000000000 +0800
+++ transmission-2.8x/libtransmission/tr-udp.c	2013-08-09 23:51:32.327381102 +0800
@@ -42,8 +42,8 @@
 /* Since we use a single UDP socket in order to implement multiple
    uTP sockets, try to set up huge buffers. */
 
-#define RECV_BUFFER_SIZE (4 * 1024 * 1024)
-#define SEND_BUFFER_SIZE (1 * 1024 * 1024)
+#define RECV_BUFFER_SIZE (2 * 1024 * 1024)
+#define SEND_BUFFER_SIZE (     512 * 1024)
 #define SMALL_BUFFER_SIZE (32 * 1024)
 
 static void
diff -ur transmission-2.82/libtransmission/tr-utp.c transmission-2.8x/libtransmission/tr-utp.c
--- transmission-2.82/libtransmission/tr-utp.c	2013-08-09 10:45:42.000000000 +0800
+++ transmission-2.8x/libtransmission/tr-utp.c	2013-08-09 23:51:32.339381103 +0800
@@ -62,7 +62,7 @@
     assert (0); /* FIXME: this is too much for the long term, but probably needed in the short term */
 }
 
-bool
+uint8_t
 UTP_Write (struct UTPSocket *socket, size_t count)
 {
     tr_logAddNamedError (MY_NAME, "UTP_RBDrained (%p, %zu) was called.", socket, count);
diff -ur transmission-2.82/Makefile.in transmission-2.8x/Makefile.in
--- transmission-2.82/Makefile.in	2013-08-09 10:49:23.000000000 +0800
+++ transmission-2.8x/Makefile.in	2013-08-09 23:51:32.351381103 +0800
@@ -451,10 +451,8 @@
 	$(SHELL) ./config.status --recheck
 
 $(top_srcdir)/configure:  $(am__configure_deps)
-	$(am__cd) $(srcdir) && $(AUTOCONF)
-$(ACLOCAL_M4):  $(am__aclocal_m4_deps)
-	$(am__cd) $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
-$(am__aclocal_m4_deps):
+	$(am__cd) $(srcdir)
+
 transmission-gtk.spec: $(top_builddir)/config.status $(srcdir)/transmission-gtk.spec.in
 	cd $(top_builddir) && $(SHELL) ./config.status $@
 qt/config.pri: $(top_builddir)/config.status $(top_srcdir)/qt/config.pri.in
