diff -ur transmission-2.71/libtransmission/rpc-server.c transmission-2.71.b/libtransmission/rpc-server.c
--- transmission-2.71/libtransmission/rpc-server.c	2012-09-27 01:02:31.374401798 +0800
+++ transmission-2.71.b/libtransmission/rpc-server.c	2012-10-04 01:02:18.512343189 +0800
@@ -358,7 +358,7 @@
 #ifdef TR_LIGHTWEIGHT
             compressionLevel = Z_DEFAULT_COMPRESSION;
 #else
-            compressionLevel = Z_BEST_COMPRESSION;
+            compressionLevel = Z_DEFAULT_COMPRESSION;
 #endif
             deflateInit2( &server->stream, compressionLevel, Z_DEFLATED, 15+16, 8, Z_DEFAULT_STRATEGY );
         }
diff -ur transmission-2.71/libtransmission/tr-udp.c transmission-2.71.b/libtransmission/tr-udp.c
--- transmission-2.71/libtransmission/tr-udp.c	2012-09-27 01:02:31.354401797 +0800
+++ transmission-2.71.b/libtransmission/tr-udp.c	2012-10-04 01:02:18.515804662 +0800
@@ -41,8 +41,8 @@
 /* Since we use a single UDP socket in order to implement multiple
    uTP sockets, try to set up huge buffers. */
 
-#define RECV_BUFFER_SIZE (4 * 1024 * 1024)
-#define SEND_BUFFER_SIZE (1 * 1024 * 1024)
+#define RECV_BUFFER_SIZE (2 * 1024 * 1024)
+#define SEND_BUFFER_SIZE (     512 * 1024)
 #define SMALL_BUFFER_SIZE (32 * 1024)
 
 static void
diff -ur transmission-2.71/libtransmission/tr-utp.c transmission-2.71.b/libtransmission/tr-utp.c
--- transmission-2.71/libtransmission/tr-utp.c	2012-09-27 01:02:31.354401797 +0800
+++ transmission-2.71.b/libtransmission/tr-utp.c	2012-10-04 03:11:42.104278107 +0800
@@ -61,7 +61,7 @@
     assert( 0 ); /* FIXME: this is too much for the long term, but probably needed in the short term */
 }
 
-bool
+uint8_t
 UTP_Write(struct UTPSocket *socket, size_t count)
 {
     tr_nerr( MY_NAME, "UTP_RBDrained(%p, %zu) was called.", socket, count );
diff -ur transmission-2.71/Makefile.in transmission-2.71.b/Makefile.in
--- transmission-2.71/Makefile.in	2012-09-27 01:03:28.446403154 +0800
+++ transmission-2.71.b/Makefile.in	2012-10-04 01:02:18.515804662 +0800
@@ -391,10 +391,8 @@
 	$(SHELL) ./config.status --recheck
 
 $(top_srcdir)/configure:  $(am__configure_deps)
-	$(am__cd) $(srcdir) && $(AUTOCONF)
-$(ACLOCAL_M4):  $(am__aclocal_m4_deps)
-	$(am__cd) $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
-$(am__aclocal_m4_deps):
+	$(am__cd) $(srcdir)
+
 transmission-gtk.spec: $(top_builddir)/config.status $(srcdir)/transmission-gtk.spec.in
 	cd $(top_builddir) && $(SHELL) ./config.status $@
 qt/config.pri: $(top_builddir)/config.status $(top_srcdir)/qt/config.pri.in
