diff -ur transmission-2.60/libtransmission/rpc-server.c transmission-2.60.b/libtransmission/rpc-server.c
--- transmission-2.60/libtransmission/rpc-server.c	2012-07-06 09:24:40.103657828 +0800
+++ transmission-2.60.b/libtransmission/rpc-server.c	2012-07-10 04:35:48.052043770 +0800
@@ -358,7 +358,7 @@
 #ifdef TR_LIGHTWEIGHT
             compressionLevel = Z_DEFAULT_COMPRESSION;
 #else
-            compressionLevel = Z_BEST_COMPRESSION;
+            compressionLevel = Z_DEFAULT_COMPRESSION;
 #endif
             deflateInit2( &server->stream, compressionLevel, Z_DEFLATED, 15+16, 8, Z_DEFAULT_STRATEGY );
         }
diff -ur transmission-2.60/libtransmission/tr-udp.c transmission-2.60.b/libtransmission/tr-udp.c
--- transmission-2.60/libtransmission/tr-udp.c	2012-07-06 09:24:40.087657827 +0800
+++ transmission-2.60.b/libtransmission/tr-udp.c	2012-07-10 04:35:48.052043770 +0800
@@ -41,8 +41,8 @@
 /* Since we use a single UDP socket in order to implement multiple
    uTP sockets, try to set up huge buffers. */
 
-#define RECV_BUFFER_SIZE (4 * 1024 * 1024)
-#define SEND_BUFFER_SIZE (1 * 1024 * 1024)
+#define RECV_BUFFER_SIZE (2 * 1024 * 1024)
+#define SEND_BUFFER_SIZE (     512 * 1024)
 #define SMALL_BUFFER_SIZE (32 * 1024)
 
 static void
Только в transmission-2.60.b/libtransmission: tr-udp.c.orig
diff -ur transmission-2.60/Makefile.in transmission-2.60.b/Makefile.in
--- transmission-2.60/Makefile.in	2012-07-06 09:27:31.203652305 +0800
+++ transmission-2.60.b/Makefile.in	2012-07-10 04:35:48.052043770 +0800
@@ -392,10 +392,8 @@
 	$(SHELL) ./config.status --recheck
 
 $(top_srcdir)/configure:  $(am__configure_deps)
-	$(am__cd) $(srcdir) && $(AUTOCONF)
-$(ACLOCAL_M4):  $(am__aclocal_m4_deps)
-	$(am__cd) $(srcdir) && $(ACLOCAL) $(ACLOCAL_AMFLAGS)
-$(am__aclocal_m4_deps):
+	$(am__cd) $(srcdir)
+
 transmission-gtk.spec: $(top_builddir)/config.status $(srcdir)/transmission-gtk.spec.in
 	cd $(top_builddir) && $(SHELL) ./config.status $@
 qt/config.pri: $(top_builddir)/config.status $(top_srcdir)/qt/config.pri.in
