diff --git a/configure.ac b/configure.ac
index df8cb50..7a67a57 100644
--- a/configure.ac
+++ b/configure.ac
@@ -455,8 +455,8 @@ AC_MSG_RESULT([$upnp_version])
 AM_CONDITIONAL([BUILD_MINIUPNP],[test "x$upnp_version" = "xnone"])
 if test "x$upnp_version" = "xnone" ; then
     LIBUPNP_CFLAGS="-I\$(top_srcdir)/third-party/"
-    LIBUPNP_LIBS="\$(top_builddir)/third-party/miniupnp/libminiupnp.a"
-    LIBUPNP_LIBS_QT="\$\${TRANSMISSION_TOP}/third-party/miniupnp/libminiupnp.a"
+    LIBUPNP_LIBS=""
+    LIBUPNP_LIBS_QT=""
 else
     AC_DEFINE([SYSTEM_MINIUPNP])
     LIBUPNP_CFLAGS="" 
@@ -475,8 +475,8 @@ dnl
 dnl  Allow usage of system natpmp library 
 
 LIBNATPMP_CFLAGS="-I\$(top_srcdir)/third-party/libnatpmp/" 
-LIBNATPMP_LIBS="\$(top_builddir)/third-party/libnatpmp/libnatpmp.a" 
-LIBNATPMP_LIBS_QT="\$\${TRANSMISSION_TOP}/third-party/libnatpmp/libnatpmp.a" 
+LIBNATPMP_LIBS="" 
+LIBNATPMP_LIBS_QT="" 
 build_bundled_natpmp="yes" 
 AC_ARG_ENABLE([external-natpmp], 
               AS_HELP_STRING([--enable-external-natpmp],[Use system external-natpmp]), 
@@ -645,7 +645,6 @@ dnl
 dnl  Generate the output
 
 AC_CONFIG_FILES([Makefile
-                 transmission-gtk.spec
                  cli/Makefile
                  daemon/Makefile
                  extras/Makefile
@@ -655,12 +654,6 @@ AC_CONFIG_FILES([Makefile
                  third-party/dht/Makefile
                  third-party/libb64/Makefile
                  third-party/libutp/Makefile
-                 third-party/libnatpmp/Makefile
-                 third-party/miniupnp/Makefile
-                 macosx/Makefile
-                 gtk/Makefile
-                 gtk/icons/Makefile
-                 qt/config.pri
                  web/Makefile
                  web/images/Makefile
                  web/style/Makefile
diff --git a/libtransmission/platform-quota.c b/libtransmission/platform-quota.c
index e7d1f6d..bb16c63 100644
--- a/libtransmission/platform-quota.c
+++ b/libtransmission/platform-quota.c
@@ -285,7 +285,7 @@ getquota (const char * device)
       spaceused = (int64_t) dq.dqb_curblocks >> 1;
 #elif defined(__APPLE__)
       spaceused = (int64_t) dq.dqb_curbytes;
-#elif defined(__UCLIBC__)
+#elif defined(__UCLIBC__) && (defined(_LINUX_QUOTA_VERSION) && _LINUX_QUOTA_VERSION < 2)
       spaceused = (int64_t) btodb(dq.dqb_curblocks);
 #elif defined(__sun) || (defined(_LINUX_QUOTA_VERSION) && _LINUX_QUOTA_VERSION < 2)
       spaceused = (int64_t) dq.dqb_curblocks >> 1;
diff --git a/libtransmission/port-forwarding.c b/libtransmission/port-forwarding.c
index 5f6c223..dbf1d70 100644
--- a/libtransmission/port-forwarding.c
+++ b/libtransmission/port-forwarding.c
@@ -64,6 +64,7 @@ getNatStateStr (int state)
 static void
 natPulse (tr_shared * s, bool do_check)
 {
+#if 0
   int oldStatus;
   int newStatus;
   tr_port public_peer_port;
@@ -90,6 +91,9 @@ natPulse (tr_shared * s, bool do_check)
     tr_logAddNamedInfo (getKey (), _("State changed from \"%1$s\" to \"%2$s\""),
              getNatStateStr (oldStatus),
              getNatStateStr (newStatus));
+#endif
+  s->natpmpStatus = TR_PORT_UNMAPPED;
+  s->upnpStatus = TR_PORT_UNMAPPED;
 }
 
 static void
@@ -180,12 +184,16 @@ stop_forwarding (tr_shared * s)
   tr_logAddNamedInfo (getKey (), "%s", _("Stopped"));
   natPulse (s, false);
 
+#if 0
   tr_natpmpClose (s->natpmp);
   s->natpmp = NULL;
+#endif
   s->natpmpStatus = TR_PORT_UNMAPPED;
 
+#if 0
   tr_upnpClose (s->upnp);
   s->upnp = NULL;
+#endif
   s->upnpStatus = TR_PORT_UNMAPPED;
 
   stop_timer (s);
diff --git a/libtransmission/session.c b/libtransmission/session.c
index bb60c23..3156ac7 100644
--- a/libtransmission/session.c
+++ b/libtransmission/session.c
@@ -343,7 +343,7 @@ tr_sessionGetDefaultSettings (tr_variant * d)
   tr_variantDictAddInt  (d, TR_KEY_peer_port_random_high,           65535);
   tr_variantDictAddStr  (d, TR_KEY_peer_socket_tos,                 TR_DEFAULT_PEER_SOCKET_TOS_STR);
   tr_variantDictAddBool (d, TR_KEY_pex_enabled,                     true);
-  tr_variantDictAddBool (d, TR_KEY_port_forwarding_enabled,         true);
+  tr_variantDictAddBool (d, TR_KEY_port_forwarding_enabled,         false);
   tr_variantDictAddInt  (d, TR_KEY_preallocation,                   TR_PREALLOCATE_SPARSE);
   tr_variantDictAddBool (d, TR_KEY_prefetch_enabled,                DEFAULT_PREFETCH_ENABLED);
   tr_variantDictAddInt  (d, TR_KEY_peer_id_ttl_hours,               6);
@@ -358,7 +358,7 @@ tr_sessionGetDefaultSettings (tr_variant * d)
   tr_variantDictAddStr  (d, TR_KEY_rpc_password,                    "");
   tr_variantDictAddStr  (d, TR_KEY_rpc_username,                    "");
   tr_variantDictAddStr  (d, TR_KEY_rpc_whitelist,                   TR_DEFAULT_RPC_WHITELIST);
-  tr_variantDictAddBool (d, TR_KEY_rpc_whitelist_enabled,           true);
+  tr_variantDictAddBool (d, TR_KEY_rpc_whitelist_enabled,           false);
   tr_variantDictAddStr  (d, TR_KEY_rpc_host_whitelist,              TR_DEFAULT_RPC_HOST_WHITELIST);
   tr_variantDictAddBool (d, TR_KEY_rpc_host_whitelist_enabled,      true);
   tr_variantDictAddInt  (d, TR_KEY_rpc_port,                        atoi (TR_DEFAULT_RPC_PORT_STR));
diff --git a/libtransmission/tr-udp.c b/libtransmission/tr-udp.c
index 2d8e6a0..9a61406 100644
--- a/libtransmission/tr-udp.c
+++ b/libtransmission/tr-udp.c
@@ -46,8 +46,8 @@ THE SOFTWARE.
 /* Since we use a single UDP socket in order to implement multiple
    uTP sockets, try to set up huge buffers. */
 
-#define RECV_BUFFER_SIZE (4 * 1024 * 1024)
-#define SEND_BUFFER_SIZE (1 * 1024 * 1024)
+#define RECV_BUFFER_SIZE (2 * 1024 * 1024)
+#define SEND_BUFFER_SIZE (512 * 1024)
 #define SMALL_BUFFER_SIZE (32 * 1024)
 
 static void
diff --git a/third-party/Makefile.am b/third-party/Makefile.am
index 24e8141..a6be52f 100644
--- a/third-party/Makefile.am
+++ b/third-party/Makefile.am
@@ -7,12 +7,6 @@ endif
 if BUILD_UTP
   UTP_DIR = libutp
 endif
-if BUILD_MINIUPNP
-  MINIUPNP_DIR = miniupnp
-endif
-if BUILD_NATPMP 
-  NATPMP_DIR = libnatpmp 
-endif 
 
 SUBDIRS = \
   $(DHT_DIR) \
