diff -urw linux-2.6.21.x/drivers/Kconfig linux-2.6.21.x.b/drivers/Kconfig
--- linux-2.6.21.x/drivers/Kconfig	2012-04-10 03:20:52.000000000 +0800
+++ linux-2.6.21.x.b/drivers/Kconfig	2012-03-20 19:16:03.000000000 +0800
@@ -2,6 +2,8 @@
 
 menu "Device Drivers"
 
+source "drivers/nvram/Kconfig"
+
 source "drivers/base/Kconfig"
 
 source "drivers/connector/Kconfig"
diff -urw linux-2.6.21.x/drivers/Makefile linux-2.6.21.x.b/drivers/Makefile
--- linux-2.6.21.x/drivers/Makefile	2012-04-10 03:20:52.000000000 +0800
+++ linux-2.6.21.x.b/drivers/Makefile	2012-03-22 15:33:35.000000000 +0800
@@ -5,6 +5,8 @@
 # Rewritten to use lists instead of if-statements.
 #
 
+obj-$(CONFIG_ASUS_NVRAM_API)	+= nvram/
+
 obj-$(CONFIG_PCI)		+= pci/
 obj-$(CONFIG_PARISC)		+= parisc/
 obj-$(CONFIG_RAPIDIO)		+= rapidio/
