diff -urw linux-2.6.21.x/init/main.c linux-2.6.21.x.b/init/main.c
--- linux-2.6.21.x/init/main.c	2012-04-26 01:38:11.076163169 +0800
+++ linux-2.6.21.x.b/init/main.c	2012-05-01 00:31:58.000000000 +0800
@@ -792,50 +792,14 @@
 #ifdef CONFIG_PROC_FS
         if (sys_mount("proc", "/proc", "proc", flags, NULL) < 0)
             printk("mount /proc file system fail!\n");
-            else
-            printk("mount /proc file system ok!\n");
 #ifdef CONFIG_USB_DEVICEFS
         if (sys_mount("usbfs", "/proc/bus/usb", "usbfs", flags, NULL) < 0)
             printk("mount /proc/bus/usb file system fail!\n");
-            else
-            printk("mount /proc/bus/usb file system ok!\n");
 #endif
 #endif
 #ifdef CONFIG_SYSFS
         if (sys_mount("sysfs", "/sys", "sysfs", flags, NULL) < 0)
             printk("mount /sys file system fail!\n");
-            else
-            printk("mount /sys file system ok!\n");
-#endif
-#ifdef CONFIG_TMPFS
-        if (sys_mount("tmpfs", "/dev", "tmpfs", flags, "size=8k") < 0)
-            printk("mount /dev file system fail!\n");
-            else
-            printk("mount /dev file system ok!\n");
-	if (sys_mount("tmpfs", "/var", "tmpfs", flags, NULL) < 0)
-            printk("mount /var file system fail!\n");
-            else
-            printk("mount /var file system ok!\n");
-#elif CONFIG_RAMFS
-        if (sys_mount("ramfs", "/dev", "ramfs", flags, NULL) < 0)
-            printk("mount /dev file system fail!\n");
-            else
-            printk("mount /dev file system ok!\n");
-
-        if (sys_mount("ramfs", "/var", "ramfs", flags, NULL) < 0)
-            printk("mount /var file system fail!\n");
-            else
-            printk("mount /var file system ok!\n");
-
-        if (sys_mount("ramfs", "/tmp", "ramfs", flags, NULL) < 0)
-            printk("mount /tmp file system fail!\n");
-            else
-            printk("mount /tmp file system ok!\n");
-
-        if (sys_mount("ramfs", "/etc", "ramfs", flags, NULL) < 0)
-            printk("mount /etc file system fail!\n");
-            else
-            printk("mount /etc file system ok!\n");
 #endif
 
 	free_initmem();
@@ -873,11 +837,12 @@
 		printk(KERN_WARNING "Failed to execute %s.  Attempting "
 					"defaults...\n", execute_command);
 	}
-#ifndef CONFIG_RT2880_ROOTFS_IN_RAM
-	run_init_process("/linuxrc");
-#endif
-	panic("No init found.  Try passing init= option to kernel.");
+	run_init_process("/sbin/init");
+	run_init_process("/etc/init");
+	run_init_process("/bin/init");
+	run_init_process("/bin/sh");
 
+	panic("No init found.  Try passing init= option to kernel.");
 	return 0;
 }
 
@@ -915,16 +880,13 @@
 
 	do_basic_setup();
 
-#ifdef CONFIG_RT2880_ROOTFS_IN_RAM
 	if (!ramdisk_execute_command)
-		ramdisk_execute_command = "/linuxrc";
-#endif
-#ifndef CONFIG_RT2880_ROOTFS_IN_RAM
+		ramdisk_execute_command = "/init";
+	
 	if (sys_access((const char __user *) ramdisk_execute_command, 0) != 0) {
 		ramdisk_execute_command = NULL;
 		prepare_namespace();
 	}
-#endif
 	/*
 	 * Ok, we have completed the initial bootup, and
 	 * we're essentially up and running. Get rid of the
