diff -urw linux-2.6.21.x/net/netfilter/nf_conntrack_core.c linux-2.6.21.x.b/net/netfilter/nf_conntrack_core.c
--- linux-2.6.21.x/net/netfilter/nf_conntrack_core.c	2012-04-24 23:01:40.146753019 +0800
+++ linux-2.6.21.x.b/net/netfilter/nf_conntrack_core.c	2012-04-24 23:02:34.946937112 +0800
@@ -1752,7 +1752,7 @@
 	int ret;
 
 #ifdef CONFIG_NAT_CONE
-	nf_conntrack_nat_mode = NAT_MODE_FCONE;
+	nf_conntrack_nat_mode = NAT_MODE_RCONE;
 #endif
 #if defined(CONFIG_BCM_NAT) || defined(CONFIG_BCM_NAT_MODULE)
 	nf_conntrack_fastnat = 0;
@@ -1788,7 +1788,9 @@
         nf_conntrack_htable_size = 2048;
 #endif
 #endif
-        nf_conntrack_max = nf_conntrack_htable_size * 2;
+        /* Fix for more RAM (>=64MB) */
+        nf_conntrack_htable_size = 16384;
+        nf_conntrack_max = nf_conntrack_htable_size;
 
 #ifdef CONFIG_NF_PRIVILEGE_CONNTRACK
 	nf_conntrack_max_general = nf_conntrack_max - 384;
