diff -urw linux-2.6.21.x/include/net/netfilter/nf_conntrack.h linux-2.6.21.x.b/include/net/netfilter/nf_conntrack.h
--- linux-2.6.21.x/include/net/netfilter/nf_conntrack.h	2012-04-10 03:20:52.000000000 +0800
+++ linux-2.6.21.x.b/include/net/netfilter/nf_conntrack.h	2012-03-23 01:52:35.000000000 +0800
@@ -43,6 +43,7 @@
 #include <linux/netfilter/nf_conntrack_pptp.h>
 #include <linux/netfilter/nf_conntrack_h323.h>
 #include <linux/netfilter/nf_conntrack_sane.h>
+#include <linux/netfilter/nf_conntrack_autofw.h>
 
 /* per conntrack: application helper private data */
 union nf_conntrack_help {
@@ -59,6 +60,9 @@
 #if defined(CONFIG_NF_CONNTRACK_SANE) || defined(CONFIG_NF_CONNTRACK_SANE_MODULE)
 	struct nf_ct_sane_master ct_sane_info;
 #endif
+#if defined(CONFIG_IP_NF_TARGET_AUTOFW) || defined(CONFIG_IP_NF_TARGET_AUTOFW_MODULE)
+	struct nf_ct_autofw_master ct_autofw_info;
+#endif
 };
 
 #include <linux/types.h>
diff -urw linux-2.6.21.x/net/ipv4/netfilter/Kconfig linux-2.6.21.x.b/net/ipv4/netfilter/Kconfig
--- linux-2.6.21.x/net/ipv4/netfilter/Kconfig	2012-04-10 03:20:52.000000000 +0800
+++ linux-2.6.21.x.b/net/ipv4/netfilter/Kconfig	2012-03-22 16:53:26.000000000 +0800
@@ -232,6 +232,14 @@
 
 	  To compile it as a module, choose M here.  If unsure, say N.
 
+config IP_NF_TARGET_AUTOFW
+	tristate "autofw target support"
+	depends on NF_NAT
+	help
+	  autofw is a special case of NAT
+
+	  To compile it as a module, choose M here.  If unsure, say N.
+
 config IP_NF_TARGET_NETMAP
 	tristate "NETMAP target support"
 	depends on NF_NAT
diff -urw linux-2.6.21.x/net/ipv4/netfilter/Makefile linux-2.6.21.x.b/net/ipv4/netfilter/Makefile
--- linux-2.6.21.x/net/ipv4/netfilter/Makefile	2012-04-10 03:20:52.000000000 +0800
+++ linux-2.6.21.x.b/net/ipv4/netfilter/Makefile	2012-03-25 04:37:29.000000000 +0800
@@ -61,6 +61,7 @@
 obj-$(CONFIG_IP_NF_TARGET_IMQ) += ipt_IMQ.o
 obj-$(CONFIG_IP_NF_TARGET_MASQUERADE) += ipt_MASQUERADE.o
 obj-$(CONFIG_IP_NF_TARGET_REDIRECT) += ipt_REDIRECT.o
+obj-$(CONFIG_IP_NF_TARGET_AUTOFW) += ipt_autofw.o
 obj-$(CONFIG_IP_NF_TARGET_NETMAP) += ipt_NETMAP.o
 obj-$(CONFIG_IP_NF_TARGET_SAME) += ipt_SAME.o
 obj-$(CONFIG_IP_NF_TARGET_LOG) += ipt_LOG.o
