diff -urw linux-2.6.21.x/drivers/usb/core/devio.c linux-2.6.21.x.b/drivers/usb/core/devio.c
--- linux-2.6.21.x/drivers/usb/core/devio.c	2012-05-02 22:40:44.764895268 +0800
+++ linux-2.6.21.x.b/drivers/usb/core/devio.c	2012-05-02 22:45:21.672582565 +0800
@@ -415,6 +415,9 @@
 	struct usb_interface *intf;
 	int err;
 
+	if (!strcmp(current->comm, "u2ec"))			// patch for U2EC
+		return 0;
+
 	if (ifnum >= 8*sizeof(ps->ifclaimed))
 		return -EINVAL;
 	/* already claimed */
