diff -urw linux-2.6.21.x/arch/mips/kernel/time.c linux-2.6.21.x.b/arch/mips/kernel/time.c
--- linux-2.6.21.x/arch/mips/kernel/time.c	2012-04-10 13:55:23.000000000 +0800
+++ linux-2.6.21.x.b/arch/mips/kernel/time.c	2012-04-18 00:34:28.000000000 +0800
@@ -59,7 +59,7 @@
  */
 static unsigned long null_rtc_get_time(void)
 {
-	return mktime(2012, 4, 24, 0, 0, 0);
+	return mktime(2000, 1, 1, 0, 0, 0);
 }
 
 static int null_rtc_set_time(unsigned long sec)
