diff -urw linux-2.6.21.x/drivers/char/ralink_gpio.c linux-2.6.21.x.b/drivers/char/ralink_gpio.c
--- linux-2.6.21.x/drivers/char/ralink_gpio.c	2012-04-10 13:55:23.000000000 +0800
+++ linux-2.6.21.x.b/drivers/char/ralink_gpio.c	2012-04-10 17:49:34.000000000 +0800
@@ -1002,7 +1002,7 @@
 	ralink_gpio_led_init_timer();
 #endif
 	printk("Ralink gpio driver initialized\n");
-	printk("Number of GPIO=%d\n",RALINK_GPIO_NUMBER);
+	printk("Number of GPIO: %d, GPIO mode: %08X\n", RALINK_GPIO_NUMBER, gpiomode);
 	return 0;
 }
