diff -urw linux-2.6.21.x/ralink/Kconfig linux-2.6.21.x.b/ralink/Kconfig
--- linux-2.6.21.x/ralink/Kconfig	2012-05-17 22:01:32.000000000 +0800
+++ linux-2.6.21.x.b/ralink/Kconfig	2012-05-18 02:24:16.625685119 +0800
@@ -16,18 +16,20 @@
 #############
 # Wire
 #############
-menu "Ethernet Ralink RT3xx/RT28xx wire driver."
+menu "Ethernet Ralink RT3xx/RT28xx wire driver"
 source "drivers/net/raeth/Kconfig"
 endmenu
 
+
 #############
 # Wireless
 #############
-menu "Platform config for  Ralink RT3xx/RT28xx/RT5xx."
+menu "Platform config for Ralink RT3xx/RT28xx/RT5xx"
 choice
         prompt "RF Type"
         depends on RALINK_RT5350
         default RALINK_RT5350_1T1R
+
         config  RALINK_RT5350_1T1R
         bool "1T1R"
 endchoice
@@ -67,9 +69,21 @@
 endchoice
 endmenu
 
-menu "Nat acell mode"
+
+# split driver in one dir
+menu "AP MODE Ralink RT3xx/RT28xx wireless driver"
+source "drivers/net/wireless/rt2860v2_ap/Kconfig"
+endmenu
+menu "STA/Client MODE Ralink RT3xx/RT28xx wireless driver"
+source "drivers/net/wireless/rt2860v2_sta/Kconfig"
+endmenu
+menu "AP MODE Ralink PCIe RT309X wireless driver"
+source "drivers/net/wireless/rt3090_ap/Kconfig"
+endmenu
+
+menu "NAT acell mode"
 choice
-        prompt "Hardware nat mode"
+        prompt "Hardware NAT mode"
 
 config  RA_NAT_HW
 	select RA_HW_NAT
@@ -87,37 +101,12 @@
                 This helps packets pass through netfilter faster when a packet
                 is an established or reply traffic.
 
-#######
-# PPE
-#######
 if RA_NAT_HW
-menu "Ralink RT3xx/RT28xx NAT Acell support."
 source "net/nat/hw_nat/Kconfig"
-endmenu
-endif
-endmenu
-
-#######
-# WIFI
-#######
-menu "AP(STA) MODE Ralink RT3xxx wireless driver."
-if RALINK_RT3052
-    source "drivers/net/wireless/ralink/rt305x/rt2860v2_ap/Kconfig"
-    source "drivers/net/wireless/ralink/rt305x/rt2860v2_sta/Kconfig"
-endif
-
-if RALINK_RT3883
-    source "drivers/net/wireless/ralink/rt3883/rt2860v2_ap/Kconfig"
-    source "drivers/net/wireless/ralink/rt3883/rt2860v2_sta/Kconfig"
-endif
-if HW_HAS_PCI
-    menu "802.11 3090 AP pci-e wireless driver."
-    source "drivers/net/wireless/ralink/rt3090/rt3090_ap/Kconfig"
-    endmenu
 endif
 endmenu
 
-menu "Ralink DEVICES."
+menu "Ralink DEVICES"
 config RTDEV_USB
 bool
 default y if RT305x_INIC_USB || RT3572_AP
diff -urw linux-2.6.21.x/drivers/net/wireless/Makefile linux-2.6.21.x.b/drivers/net/wireless/Makefile
--- linux-2.6.21.x/drivers/net/wireless/Makefile	2012-05-17 22:01:26.000000000 +0800
+++ linux-2.6.21.x.b/drivers/net/wireless/Makefile	2012-05-10 22:01:13.382710315 +0800
@@ -44,12 +44,6 @@
 
 obj-$(CONFIG_USB_ZD1201)	+= zd1201.o
 
-ifdef CONFIG_RALINK_RT3052
-obj-$(CONFIG_RT2860V2_AP)       += ralink/rt305x/rt2860v2_ap/
-obj-$(CONFIG_RT2860V2_STA)      += ralink/rt305x/rt2860v2_sta/
-endif
-ifdef CONFIG_RALINK_RT3883
-obj-$(CONFIG_RT2860V2_AP)       += ralink/rt3883/rt2860v2_ap/
-obj-$(CONFIG_RT2860V2_STA)      += ralink/rt3883/rt2860v2_sta/
-endif
-obj-$(CONFIG_RT3090_AP)         += ralink/rt3090/rt3090_ap/
+obj-$(CONFIG_RT2860V2_AP)       += rt2860v2_ap/
+obj-$(CONFIG_RT2860V2_STA)      += rt2860v2_sta/
+obj-$(CONFIG_RT3090_AP)         += rt3090_ap/
