diff -urw linux-2.6.21.x/net/core/dev.c linux-2.6.21.x.b/net/core/dev.c
--- linux-2.6.21.x/net/core/dev.c	2012-06-05 21:11:43.000000000 +0800
+++ linux-2.6.21.x.b/net/core/dev.c	2012-06-11 19:03:38.194808382 +0800
@@ -138,7 +138,7 @@
 #endif
 #endif
 
-#if defined (CONFIG_RA_HW_NAT)  || defined (CONFIG_RA_HW_NAT_MODULE)
+#if defined(CONFIG_RA_HW_NAT_PCI) && (defined(CONFIG_RA_HW_NAT) || defined(CONFIG_RA_HW_NAT_MODULE))
 #include "../net/nat/hw_nat/ra_nat.h"
 extern void (*ra_sw_nat_hook_rs) (uint32_t Ebl);
 #endif
@@ -907,8 +907,8 @@
 		 */
 		raw_notifier_call_chain(&netdev_chain, NETDEV_UP, dev);
 
-#if defined(CONFIG_RA_HW_NAT) || defined(CONFIG_RA_HW_NAT_MODULE)
-		if((strncmp(dev->name, "eth", 3) == 0) && (ra_sw_nat_hook_rs != NULL)) {
+#if defined(CONFIG_RA_HW_NAT_PCI) && (defined(CONFIG_RA_HW_NAT) || defined(CONFIG_RA_HW_NAT_MODULE))
+		if((strcmp(dev->name, "eth0") == 0) && (ra_sw_nat_hook_rs != NULL)) {
 		    /* reconfigure dstif table in hw_nat module */
 		    ra_sw_nat_hook_rs(0);
 		    ra_sw_nat_hook_rs(1);
