diff -ur dnsmasq-2.64/CHANGELOG dnsmasq-2.64.b/CHANGELOG
--- dnsmasq-2.64/CHANGELOG	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/CHANGELOG	2012-12-09 04:29:13.377776136 +0800
@@ -1,3 +1,8 @@
+version 2.65
+	    Fix regression which broke forwarding of queries sent via
+	    TCP which are not for A and AAAA and which were directed to
+	    non-default servers. Thanks to Niax for the bug report.
+
 version 2.64
             Handle DHCP FQDN options with all flag bits zero and
             --dhcp-client-update set. Thanks to Bernd Krumbroeck for
diff -ur dnsmasq-2.64/Makefile dnsmasq-2.64.b/Makefile
--- dnsmasq-2.64/Makefile	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/Makefile	2012-12-09 04:41:18.481766332 +0800
@@ -18,7 +18,7 @@
 
 # Variables you may well want to override.
 
-PREFIX        = /usr/local
+PREFIX        = /usr
 BINDIR        = $(PREFIX)/sbin
 MANDIR        = $(PREFIX)/share/man
 LOCALEDIR     = $(PREFIX)/share/locale
diff -ur dnsmasq-2.64/src/config.h dnsmasq-2.64.b/src/config.h
--- dnsmasq-2.64/src/config.h	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/config.h	2012-12-09 04:17:36.389785557 +0800
@@ -118,9 +118,9 @@
 #define HAVE_DHCP
 #define HAVE_DHCP6 
 #define HAVE_TFTP
-#define HAVE_SCRIPT
+/* #define HAVE_SCRIPT */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -206,10 +206,12 @@
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#elif !defined(NO_IPV6)
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
diff -ur dnsmasq-2.64/src/forward.c dnsmasq-2.64.b/src/forward.c
--- dnsmasq-2.64/src/forward.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/forward.c	2012-12-09 04:27:41.465777377 +0800
@@ -834,7 +834,8 @@
   int norebind = 0;
   int checking_disabled;
   size_t m;
-  unsigned short qtype, gotname;
+  unsigned short qtype;
+  unsigned int gotname;
   unsigned char c1, c2;
   /* Max TCP packet + slop */
   unsigned char *packet = whine_malloc(65536 + MAXDNAME + RRFIXEDSZ);
diff -ur dnsmasq-2.64/src/lease.c dnsmasq-2.64.b/src/lease.c
--- dnsmasq-2.64/src/lease.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/lease.c	2012-12-09 04:15:19.161787414 +0800
@@ -812,8 +812,11 @@
   struct dhcp_lease *lease_tmp;
   char *new_name = NULL, *new_fqdn = NULL;
 
+#if 0
+  /* disable warning, noisy */
   if (config_domain && (!domain || !hostname_isequal(domain, config_domain)))
     my_syslog(MS_DHCP | LOG_WARNING, _("Ignoring domain %s for DHCP host name %s"), config_domain, name);
+#endif
   
   if (lease->hostname && name && hostname_isequal(lease->hostname, name))
     {
diff -ur dnsmasq-2.64/src/option.c dnsmasq-2.64.b/src/option.c
--- dnsmasq-2.64/src/option.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/option.c	2012-12-09 04:30:36.449775011 +0800
@@ -3002,7 +3002,7 @@
     case LOPT_RR: /* dns-rr */
       {
        	struct txt_record *new;
-	size_t len;
+	size_t len = 0;
 	char *data;
 	int val;
 
@@ -3019,12 +3019,11 @@
 	  ret_err(_("bad RR record"));
 	   	
 	new->class = val;
-	new->len = 0;
+	new->len = len;
 	
-	if (data)
+	if (data && len)
 	  {
 	    new->txt=opt_malloc(len);
-	    new->len = len;
 	    memcpy(new->txt, data, len);
 	  }
 	
diff -ur dnsmasq-2.64/src/radv.c dnsmasq-2.64.b/src/radv.c
--- dnsmasq-2.64/src/radv.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/radv.c	2012-12-09 04:15:19.161787414 +0800
@@ -172,7 +172,10 @@
 	  mac = daemon->namebuff;
 	}
          
+#if 0
+/* disable Router Advertisement messages, noisy */
       my_syslog(MS_DHCP | LOG_INFO, "RTR-SOLICIT(%s) %s", interface, mac);
+#endif
       /* source address may not be valid in solicit request. */
       send_ra(if_index, interface, !IN6_IS_ADDR_UNSPECIFIED(&from.sin6_addr) ? &from.sin6_addr : NULL);
     }
@@ -420,7 +423,10 @@
 		  opt->prefix = *local;
 		  
 		  inet_ntop(AF_INET6, local, daemon->addrbuff, ADDRSTRLEN);
+#if 0
+/* disable Router Advertisement messages, noisy */
 		  my_syslog(MS_DHCP | LOG_INFO, "RTR-ADVERT(%s) %s", param->if_name, daemon->addrbuff); 		    
+#endif
 		}
 	    }
 	}
diff -ur dnsmasq-2.64/src/rfc2131.c dnsmasq-2.64.b/src/rfc2131.c
--- dnsmasq-2.64/src/rfc2131.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/rfc2131.c	2012-12-09 04:15:19.161787414 +0800
@@ -1505,6 +1505,8 @@
 static void log_packet(char *type, void *addr, unsigned char *ext_mac, 
 		       int mac_len, char *interface, char *string, u32 xid)
 {
+#if 0
+/* disable noisy DHCP server syslog */
   struct in_addr a;
  
   /* addr may be misaligned */
@@ -1530,6 +1532,7 @@
 	      addr ? " " : "",
 	      daemon->namebuff,
 	      string ? string : "");
+#endif
 }
 
 static void log_options(unsigned char *start, u32 xid)
diff -ur dnsmasq-2.64/src/rfc3315.c dnsmasq-2.64.b/src/rfc3315.c
--- dnsmasq-2.64/src/rfc3315.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/rfc3315.c	2012-12-09 04:15:19.161787414 +0800
@@ -106,8 +106,11 @@
 
       if (!context)
 	{
+#if 0
+/* disable DHCPv6 noaddr messages, noisy */
 	  my_syslog(MS_DHCP | LOG_WARNING, 
 		    _("no address range available for DHCPv6 request via %s"), iface_name);
+#endif
 	  return 0;
 	}
 
@@ -1428,6 +1431,8 @@
  
 static void log6_packet(char *type, unsigned char *clid, int clid_len, struct in6_addr *addr, int xid, char *iface, char *string)
 {
+#if 0
+/* disable noisy DHCPv6 server syslog */
   /* avoid buffer overflow */
   if (clid_len > 100)
     clid_len = 100;
@@ -1457,6 +1462,7 @@
 	      daemon->dhcp_buff2,
 	      daemon->namebuff,
 	      string ? string : "");
+#endif
 }
 
 static void *opt6_find (void *opts, void *end, unsigned int search, unsigned int minsize)
diff -ur dnsmasq-2.64/src/slaac.c dnsmasq-2.64.b/src/slaac.c
--- dnsmasq-2.64/src/slaac.c	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/src/slaac.c	2012-12-09 04:15:19.161787414 +0800
@@ -205,7 +205,10 @@
 	    slaac->backoff = 0;
 	    gotone = 1;
 	    inet_ntop(AF_INET6, sender, daemon->addrbuff, ADDRSTRLEN);
+#if 0
+/* disable DHCPv6 SLAAC messages, noisy */
 	    my_syslog(MS_DHCP | LOG_INFO, "SLAAC-CONFIRM(%s) %s %s", interface, daemon->addrbuff, lease->hostname); 
+#endif
 	  }
   
   lease_update_dns(gotone);
diff -ur dnsmasq-2.64/VERSION dnsmasq-2.64.b/VERSION
--- dnsmasq-2.64/VERSION	2012-12-03 22:05:59.000000000 +0800
+++ dnsmasq-2.64.b/VERSION	2012-12-09 04:29:18.933776063 +0800
@@ -1 +1 @@
- (HEAD, v2.64rc3, v2.64, master)
+ (HEAD, v2.64+, master)
