diff -ur dnsmasq-2.66/Makefile dnsmasq-2.66.b/Makefile
--- dnsmasq-2.66/Makefile	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/Makefile	2013-04-21 01:27:56.064720205 +0800
@@ -18,14 +18,14 @@
 
 # Variables you may well want to override.
 
-PREFIX        = /usr/local
+PREFIX        = /usr
 BINDIR        = $(PREFIX)/sbin
 MANDIR        = $(PREFIX)/share/man
 LOCALEDIR     = $(PREFIX)/share/locale
 BUILDDIR      = $(SRC)
 DESTDIR       = 
-CFLAGS        = -Wall -W -O2
-LDFLAGS       = 
+CFLAGS        = -Wall -W -O2 -ffunction-sections -fdata-sections
+LDFLAGS       = -Wl,--gc-sections
 COPTS         = 
 RPM_OPT_FLAGS = 
 LIBS          = 
diff -ur dnsmasq-2.66/src/config.h dnsmasq-2.66.b/src/config.h
--- dnsmasq-2.66/src/config.h	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/config.h	2013-04-21 01:30:15.280803479 +0800
@@ -133,11 +133,11 @@
 #define HAVE_DHCP
 #define HAVE_DHCP6 
 #define HAVE_TFTP
-#define HAVE_SCRIPT
-#define HAVE_AUTH
-#define HAVE_IPSET 
+/* #define HAVE_SCRIPT */
+/* #define HAVE_AUTH */
+/* #define HAVE_IPSET */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -222,10 +222,12 @@
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#elif !defined(NO_IPV6)
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
diff -ur dnsmasq-2.66/src/dnsmasq.c dnsmasq-2.66.b/src/dnsmasq.c
--- dnsmasq-2.66/src/dnsmasq.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/dnsmasq.c	2013-04-21 01:27:56.064720205 +0800
@@ -566,6 +566,8 @@
 #endif
 
 #ifdef HAVE_TFTP
+#if 0
+/* disable exit on tftp dir not exist (dir may be mounted later) */
       if (option_bool(OPT_TFTP))
     {
       DIR *dir;
@@ -592,6 +594,7 @@
 	}
     }
 #endif
+#endif
 
   if (daemon->port == 0)
     my_syslog(LOG_INFO, _("started, version %s DNS disabled"), VERSION);
diff -ur dnsmasq-2.66/src/lease.c dnsmasq-2.66.b/src/lease.c
--- dnsmasq-2.66/src/lease.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/lease.c	2013-04-21 01:27:56.064720205 +0800
@@ -848,8 +848,11 @@
   struct dhcp_lease *lease_tmp;
   char *new_name = NULL, *new_fqdn = NULL;
 
+#if 0
+  /* disable warning, noisy */
   if (config_domain && (!domain || !hostname_isequal(domain, config_domain)))
     my_syslog(MS_DHCP | LOG_WARNING, _("Ignoring domain %s for DHCP host name %s"), config_domain, name);
+#endif
   
   if (lease->hostname && name && hostname_isequal(lease->hostname, name))
     {
diff -ur dnsmasq-2.66/src/option.c dnsmasq-2.66.b/src/option.c
--- dnsmasq-2.66/src/option.c	2013-04-19 17:22:06.000000000 +0800
+++ dnsmasq-2.66.b/src/option.c	2013-04-21 01:27:56.064720205 +0800
@@ -3287,7 +3287,7 @@
     case LOPT_RR: /* dns-rr */
       {
        	struct txt_record *new;
-	size_t len;
+	size_t len = 0;
 	char *data;
 	int val;
 
@@ -3304,12 +3304,11 @@
 	  ret_err(_("bad RR record"));
 	   	
 	new->class = val;
-	new->len = 0;
+	new->len = len;
 	
-	if (data)
+	if (data && len)
 	  {
 	    new->txt=opt_malloc(len);
-	    new->len = len;
 	    memcpy(new->txt, data, len);
 	  }
 	
diff -ur dnsmasq-2.66/src/radv.c dnsmasq-2.66.b/src/radv.c
--- dnsmasq-2.66/src/radv.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/radv.c	2013-04-21 01:27:56.068725673 +0800
@@ -177,7 +177,10 @@
 	  mac = daemon->namebuff;
 	}
          
+#if 0
+/* disable Router Advertisement messages, noisy */
       my_syslog(MS_DHCP | LOG_INFO, "RTR-SOLICIT(%s) %s", interface, mac);
+#endif
       /* source address may not be valid in solicit request. */
       send_ra(now, if_index, interface, !IN6_IS_ADDR_UNSPECIFIED(&from.sin6_addr) ? &from.sin6_addr : NULL);
     }
@@ -464,7 +467,10 @@
 		  opt->prefix = *local;
 		  
 		  inet_ntop(AF_INET6, local, daemon->addrbuff, ADDRSTRLEN);
+#if 0
+/* disable Router Advertisement messages, noisy */
 		  my_syslog(MS_DHCP | LOG_INFO, "RTR-ADVERT(%s) %s", param->if_name, daemon->addrbuff); 		    
+#endif
 		}
 	    }
 	}
diff -ur dnsmasq-2.66/src/rfc2131.c dnsmasq-2.66.b/src/rfc2131.c
--- dnsmasq-2.66/src/rfc2131.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/rfc2131.c	2013-04-21 01:27:56.068725673 +0800
@@ -1518,6 +1518,8 @@
 static void log_packet(char *type, void *addr, unsigned char *ext_mac, 
 		       int mac_len, char *interface, char *string, u32 xid)
 {
+#if 0
+/* disable noisy DHCP server syslog */
   struct in_addr a;
  
   /* addr may be misaligned */
@@ -1543,6 +1545,7 @@
 	      addr ? " " : "",
 	      daemon->namebuff,
 	      string ? string : "");
+#endif
 }
 
 static void log_options(unsigned char *start, u32 xid)
diff -ur dnsmasq-2.66/src/rfc3315.c dnsmasq-2.66.b/src/rfc3315.c
--- dnsmasq-2.66/src/rfc3315.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/rfc3315.c	2013-04-21 01:45:00.681220771 +0800
@@ -141,8 +141,11 @@
 
       if (!context)
 	{
+#if 0
+/* disable DHCPv6 noaddr messages, noisy */
 	  my_syslog(MS_DHCP | LOG_WARNING, 
 		    _("no address range available for DHCPv6 request via %s"), iface_name);
+#endif
 	  return 0;
 	}
 
@@ -1739,6 +1742,9 @@
  
 static void log6_packet(struct state *state, char *type, struct in6_addr *addr, char *string)
 {
+#if 0
+/* disable noisy DHCPv6 server syslog */
+
   int clid_len = state->clid_len;
 
   /* avoid buffer overflow */
@@ -1770,6 +1776,7 @@
 	      daemon->dhcp_buff2,
 	      daemon->namebuff,
 	      string ? string : "");
+#endif
 }
 
 static void *opt6_find (void *opts, void *end, unsigned int search, unsigned int minsize)
diff -ur dnsmasq-2.66/src/slaac.c dnsmasq-2.66.b/src/slaac.c
--- dnsmasq-2.66/src/slaac.c	2013-04-17 20:52:49.000000000 +0800
+++ dnsmasq-2.66.b/src/slaac.c	2013-04-21 01:27:56.068725673 +0800
@@ -198,7 +198,10 @@
 	    slaac->backoff = 0;
 	    gotone = 1;
 	    inet_ntop(AF_INET6, sender, daemon->addrbuff, ADDRSTRLEN);
+#if 0
+/* disable DHCPv6 SLAAC messages, noisy */
 	    my_syslog(MS_DHCP | LOG_INFO, "SLAAC-CONFIRM(%s) %s %s", interface, daemon->addrbuff, lease->hostname); 
+#endif
 	  }
   
   lease_update_dns(gotone);
