diff -urN dnsmasq-2.68/Makefile dnsmasq-2.68.b/Makefile
--- dnsmasq-2.68/Makefile	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/Makefile	2013-12-11 04:23:22.499996373 +0800
@@ -18,14 +18,14 @@
 
 # Variables you may well want to override.
 
-PREFIX        = /usr/local
+PREFIX        = /usr
 BINDIR        = $(PREFIX)/sbin
 MANDIR        = $(PREFIX)/share/man
 LOCALEDIR     = $(PREFIX)/share/locale
 BUILDDIR      = $(SRC)
 DESTDIR       = 
-CFLAGS        = -Wall -W -O2
-LDFLAGS       = 
+CFLAGS        = -Wall -W -O2 -ffunction-sections -fdata-sections
+LDFLAGS       = -Wl,--gc-sections
 COPTS         = 
 RPM_OPT_FLAGS = 
 LIBS          = 
diff -urN dnsmasq-2.68/src/config.h dnsmasq-2.68.b/src/config.h
--- dnsmasq-2.68/src/config.h	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/src/config.h	2013-12-11 04:23:22.499996373 +0800
@@ -131,11 +131,11 @@
 #define HAVE_DHCP
 #define HAVE_DHCP6 
 #define HAVE_TFTP
-#define HAVE_SCRIPT
-#define HAVE_AUTH
-#define HAVE_IPSET 
+/* #define HAVE_SCRIPT */
+/* #define HAVE_AUTH */
+/* #define HAVE_IPSET */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -214,10 +214,12 @@
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#elif !defined(NO_IPV6)
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
diff -urN dnsmasq-2.68/src/dnsmasq.c dnsmasq-2.68.b/src/dnsmasq.c
--- dnsmasq-2.68/src/dnsmasq.c	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/src/dnsmasq.c	2013-12-11 04:23:22.499996373 +0800
@@ -583,6 +583,8 @@
 #endif
 
 #ifdef HAVE_TFTP
+#if 0
+/* disable exit on tftp dir not exist (dir may be mounted later) */
       if (option_bool(OPT_TFTP))
     {
       DIR *dir;
@@ -609,6 +611,7 @@
 	}
     }
 #endif
+#endif
 
   if (daemon->port == 0)
     my_syslog(LOG_INFO, _("started, version %s DNS disabled"), VERSION);
diff -urN dnsmasq-2.68/src/lease.c dnsmasq-2.68.b/src/lease.c
--- dnsmasq-2.68/src/lease.c	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/src/lease.c	2013-12-11 04:23:22.499996373 +0800
@@ -873,8 +873,11 @@
   struct dhcp_lease *lease_tmp;
   char *new_name = NULL, *new_fqdn = NULL;
 
+#if 0
+  /* disable warning, noisy */
   if (config_domain && (!domain || !hostname_isequal(domain, config_domain)))
     my_syslog(MS_DHCP | LOG_WARNING, _("Ignoring domain %s for DHCP host name %s"), config_domain, name);
+#endif
   
   if (lease->hostname && name && hostname_isequal(lease->hostname, name))
     {
diff -urN dnsmasq-2.68/src/network.c dnsmasq-2.68.b/src/network.c
--- dnsmasq-2.68/src/network.c	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/src/network.c	2013-12-11 04:23:22.499996373 +0800
@@ -1286,7 +1286,9 @@
 	      break;
 	  if (iface)
 	    {
+#if 0
 	      my_syslog(LOG_WARNING, _("ignoring nameserver %s - local interface"), daemon->namebuff);
+#endif
 	      free(new);
 	      continue;
 	    }
diff -urN dnsmasq-2.68/src/rfc3315.c dnsmasq-2.68.b/src/rfc3315.c
--- dnsmasq-2.68/src/rfc3315.c	2013-12-08 23:58:29.000000000 +0800
+++ dnsmasq-2.68.b/src/rfc3315.c	2013-12-11 04:23:22.504801313 +0800
@@ -159,8 +159,11 @@
 	  
       if (!state->context)
 	{
+#if 0
+/* disable DHCPv6 noaddr messages, noisy */
 	  my_syslog(MS_DHCP | LOG_WARNING, 
 		    _("no address range available for DHCPv6 request via %s"), state->iface_name);
+#endif
 	  return 0;
 	}
 
@@ -802,7 +805,7 @@
 	    put_opt6_short(DHCP6NOADDRS);
 	    put_opt6_string(_("no addresses available"));
 	    end_opt6(o1);
-	    log6_packet(state, "DHCPADVERTISE", NULL, _("no addresses available"));
+	    log6_quiet(state, "DHCPADVERTISE", NULL, _("no addresses available"));
 	  }
 
 	break;
@@ -917,7 +920,7 @@
 	    put_opt6_short(DHCP6NOADDRS);
 	    put_opt6_string(_("no addresses available"));
 	    end_opt6(o1);
-	    log6_packet(state, "DHCPREPLY", NULL, _("no addresses available"));
+	    log6_quiet(state, "DHCPREPLY", NULL, _("no addresses available"));
 	  }
 
 	tagif = add_options(state, 0);
