diff -ur dnsmasq-2.62/Makefile dnsmasq-2.62.b/Makefile
--- dnsmasq-2.62/Makefile	2012-06-05 04:40:11.000000000 +0800
+++ dnsmasq-2.62.b/Makefile	2012-07-15 03:14:54.022690310 +0800
@@ -18,7 +18,7 @@
 
 # Variables you may well want to override.
 
-PREFIX        = /usr/local
+PREFIX        = /usr
 BINDIR        = $(PREFIX)/sbin
 MANDIR        = $(PREFIX)/share/man
 LOCALEDIR     = $(PREFIX)/share/locale
diff -ur dnsmasq-2.62/src/config.h dnsmasq-2.62.b/src/config.h
--- dnsmasq-2.62/src/config.h	2012-06-05 04:40:11.000000000 +0800
+++ dnsmasq-2.62.b/src/config.h	2012-07-15 03:15:13.244867523 +0800
@@ -117,11 +117,11 @@
    has no library dependencies other than libc */
 
 #define HAVE_DHCP
-#define HAVE_DHCP6 
+/* #define HAVE_DHCP6 */
 #define HAVE_TFTP
-#define HAVE_SCRIPT
+/* #define HAVE_SCRIPT */
 /* #define HAVE_LUASCRIPT */
-/* #define HAVE_BROKEN_RTC */
+#define HAVE_BROKEN_RTC
 /* #define HAVE_DBUS */
 /* #define HAVE_IDN */
 /* #define HAVE_CONNTRACK */
@@ -201,10 +201,13 @@
 #if !defined(__ARCH_HAS_MMU__) && !defined(__UCLIBC_HAS_MMU__)
 #  define NO_FORK
 #endif
-#if defined(__UCLIBC_HAS_IPV6__)
+#if defined(__UCLIBC_HAS_IPV6__) && defined(USE_IPV6)
 #  ifndef IPV6_V6ONLY
 #    define IPV6_V6ONLY 26
 #  endif
+#else
+#  undef HAVE_IPV6
+#  define NO_IPV6
 #endif
 
 /* This is for glibc 2.x */
diff -ur dnsmasq-2.62/src/lease.c dnsmasq-2.62.b/src/lease.c
--- dnsmasq-2.62/src/lease.c	2012-06-05 04:40:11.000000000 +0800
+++ dnsmasq-2.62.b/src/lease.c	2012-07-15 03:14:54.026723225 +0800
@@ -799,9 +799,12 @@
   struct dhcp_lease *lease_tmp;
   char *new_name = NULL, *new_fqdn = NULL;
 
+#if 0
+  /* disable warning, noisy */
   if (config_domain && (!domain || !hostname_isequal(domain, config_domain)))
     my_syslog(MS_DHCP | LOG_WARNING, _("Ignoring domain %s for DHCP host name %s"), config_domain, name);
-  
+#endif
+
   if (lease->hostname && name && hostname_isequal(lease->hostname, name))
     {
       if (auth)
diff -ur dnsmasq-2.62/src/rfc2131.c dnsmasq-2.62.b/src/rfc2131.c
--- dnsmasq-2.62/src/rfc2131.c	2012-06-05 04:40:11.000000000 +0800
+++ dnsmasq-2.62.b/src/rfc2131.c	2012-07-15 03:14:54.026723225 +0800
@@ -1502,6 +1502,8 @@
 static void log_packet(char *type, void *addr, unsigned char *ext_mac, 
 		       int mac_len, char *interface, char *string, u32 xid)
 {
+#if 0
+/* disable noisy DHCP server syslog */
   struct in_addr a;
  
   /* addr may be misaligned */
@@ -1527,6 +1529,7 @@
 	      addr ? " " : "",
 	      daemon->namebuff,
 	      string ? string : "");
+#endif
 }
 
 static void log_options(unsigned char *start, u32 xid)
