diff -ur curl-7.63.0-ref/Makefile.am curl-7.63.0/Makefile.am
--- curl-7.63.0-ref/Makefile.am	2018-10-27 17:11:01.000000000 +0300
+++ curl-7.63.0/Makefile.am	2018-12-28 19:42:44.541739033 +0300
@@ -168,7 +168,7 @@
 bin_SCRIPTS = curl-config
 
 SUBDIRS = lib src
-DIST_SUBDIRS = $(SUBDIRS) tests packages scripts include docs
+DIST_SUBDIRS = $(SUBDIRS) packages include
 
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
@@ -279,8 +279,6 @@
 # We extend the standard install with a custom hook:
 install-data-hook:
 	cd include && $(MAKE) install
-	cd docs && $(MAKE) install
-	cd docs/libcurl && $(MAKE) install
 
 # We extend the standard uninstall with a custom hook:
 uninstall-hook:
diff -ur curl-7.63.0-ref/Makefile.in curl-7.63.0/Makefile.in
--- curl-7.63.0-ref/Makefile.in	2018-12-12 10:14:23.000000000 +0300
+++ curl-7.63.0/Makefile.in	2018-12-28 19:43:30.224562122 +0300
@@ -644,7 +644,7 @@
 
 bin_SCRIPTS = curl-config
 SUBDIRS = lib src
-DIST_SUBDIRS = $(SUBDIRS) tests packages scripts include docs
+DIST_SUBDIRS = $(SUBDIRS) packages include
 pkgconfigdir = $(libdir)/pkgconfig
 pkgconfig_DATA = libcurl.pc
 LIB_VAUTH_CFILES = vauth/vauth.c vauth/cleartext.c vauth/cram.c         \
@@ -1449,8 +1449,6 @@
 # We extend the standard install with a custom hook:
 install-data-hook:
 	cd include && $(MAKE) install
-	cd docs && $(MAKE) install
-	cd docs/libcurl && $(MAKE) install
 
 # We extend the standard uninstall with a custom hook:
 uninstall-hook:
