--- a/Makefile.am
+++ b/Makefile.am
@@ -1,4 +1,4 @@
-SUBDIRS =  po lib deps src doc test
+SUBDIRS =  po lib deps src
 
 ACLOCAL_AMFLAGS = -I m4 --install
 RST2HTML = @RST2HTML@
--- a/Makefile.in
+++ b/Makefile.in
@@ -449,7 +449,7 @@ target_vendor = @target_vendor@
 top_build_prefix = @top_build_prefix@
 top_builddir = @top_builddir@
 top_srcdir = @top_srcdir@
-SUBDIRS = po lib deps src doc test
+SUBDIRS = po lib deps src
 ACLOCAL_AMFLAGS = -I m4 --install
 EXTRA_DIST = config.rpath \
 	     script-helper \
