---
 win32/Makefile.gcc |    8 ++++++--
 1 file changed, 6 insertions(+), 2 deletions(-)

--- a/win32/Makefile.gcc
+++ b/win32/Makefile.gcc
@@ -71,7 +71,11 @@
        gzwrite.o infback.o inffast.o inflate.o inftrees.o trees.o uncompr.o zutil.o
 OBJA =
 
-all: $(STATICLIB) $(SHAREDLIB) $(IMPLIB) example.exe minigzip.exe example_d.exe minigzip_d.exe
+all: $(STATICLIB) example.exe minigzip.exe
+
+ifeq ($(SHARED_MODE),1)
+all: $(SHAREDLIB) $(IMPLIB) example_d.exe minigzip_d.exe
+endif
 
 test: example.exe minigzip.exe
 	./example
@@ -125,7 +129,7 @@
 .PHONY: install uninstall clean
 
 install: zlib.h zconf.h $(STATICLIB) $(IMPLIB)
-	@if test -z "$(DESTDIR)$(INCLUDE_PATH)" -o -z "$(DESTDIR)$(LIBRARY_PATH)" -o -z "$(DESTDIR)$(BINARY_PATH)"; then \
+	@if test -z "$(INCLUDE_PATH)" -o -z "$(LIBRARY_PATH)" -o -z "$(BINARY_PATH)"; then \
 		echo INCLUDE_PATH, LIBRARY_PATH, and BINARY_PATH must be specified; \
 		exit 1; \
 	fi
