--- njit8021xclient-1.3.1-pdv/src/md5-buildin/md5_dgst.c	2020-03-07 23:07:35.000000000 +0800
+++ njit8021xclient-1.3.1-pdv/src/md5-buildin/md5_dgst.c	2020-03-08 23:07:35.000000000 +0800
@@ -415,7 +415,7 @@
 	a=ROTATE(a,s); \
 	a+=b; };
 
-const char MD5_version[]="MD5" OPENSSL_VERSION_PTEXT;
+const char MD5_version[]="MD5" OPENSSL_VERSION_TEXT;
 
 /* Implemented from RFC1321 The MD5 Message-Digest Algorithm
  */
