From 0b2586e1428545ece4998b97613e139a50723945 Mon Sep 17 00:00:00 2001
From: Mike Frysinger <vapier@gentoo.org>
Date: Sat, 19 Apr 2008 05:28:39 +0000
Subject: update AI flags from trunk (makes newer GNU apps happy)

---
diff --git a/include/netdb.h b/include/netdb.h
index 4a48d0b..bed5953 100644
--- a/include/netdb.h
+++ b/include/netdb.h
@@ -401,6 +401,9 @@ struct addrinfo
 # define AI_PASSIVE	0x0001	/* Socket address is intended for `bind'.  */
 # define AI_CANONNAME	0x0002	/* Request for canonical name.  */
 # define AI_NUMERICHOST	0x0004	/* Don't use name resolution.  */
+# define AI_V4MAPPED	0x0008  /* IPv4 mapped addresses are acceptable.  */
+# define AI_ALL		0x0010	/* Return IPv4 mapped and IPv6 addresses.  */
+# define AI_ADDRCONFIG	0x0020	/* Use host config to choose ret addr type. */
 
 /* Error values for `getaddrinfo' function.  */
 # define EAI_BADFLAGS	  -1	/* Invalid value for `ai_flags' field.  */
--
cgit v0.9.0.1-2-gef13
